lib/optimizely/config_manager/polling_project_config_manager.rb

```diff
@@ -0,0 +1,177 @@
+    end
+
+    def get_datafile_url(sdk_key, url, url_template)
+      # Helper method to determine URL from where to fetch the datafile.
+      # sdk_key - Key uniquely identifying the datafile.
+      # url - String representing URL from which to fetch the datafile.
+      # url_template - String representing template which is filled in with
+      #               SDK key to determine URL from which to fetch the datafile.
+      # Returns String representing URL to fetch datafile from.
+
+      raise InvalidInputsError, 'Must provide at least one of sdk_key or url.' if sdk_key.nil? && url.nil?
+
+      return (url_template % sdk_key) unless url
+
+      url
+    end
+
+    def update_interval(update_interval)
+      # Helper method to set frequency at which datafile has to be polled and ProjectConfig updated.
+      #
+      # update_interval - Time in seconds after which to update datafile.
+
+      # If polling interval is less than minimum allowed interval then set it to default update interval.
+
+      return unless @update_interval < Helpers::Constants::CONFIG_MANAGER['MIN_UPDATE_INTERVAL']
+
+      @logger.log(
+        Logger::DEBUG,
+        "Invalid update_interval #{update_interval} provided. Defaulting to #{Helpers::Constants::CONFIG_MANAGER['DEFAULT_UPDATE_INTERVAL']}"
+      )
+      @update_interval = Helpers::Constants::CONFIG_MANAGER['DEFAULT_UPDATE_INTERVAL']
+    end
+
+    def handle_response(response)
+      # Helper method to handle response containing datafile.
+      #
+      # response - requests.Response
+
+      # Leave datafile and config unchanged if it has not been modified.
+      if response.code == Net::HTTPNotModified
+        @logger.log(
+          Logger::DEBUG,
+          "Not updating config as datafile has not updated since #{@last_modified}."
+        )
+        return
+      end
+
+      @last_modified = response[Helpers::Constants::HTTP_HEADERS['LAST_MODIFIED']]
+      @config = set_config response.body
+    end
+
+    def fetch_datafile
+      # Fetch datafile and set ProjectConfig.
+      begin
+        uri = URI(@datafile_url)
+        request = Net::HTTP::Get.new(uri)
+        request[Helpers::Constants::HTTP_HEADERS['IF_MODIFIED_SINCE']] = @last_modified if @last_modified
+        response = Net::HTTP.start(uri.hostname, uri.port, read_timeout: Helpers::Constants::CONFIG_MANAGER['REQUEST_TIMEOUT']) do |http|
+          http.request(request)
+        end
+      rescue StandardError => e
+        @logger.log(
+          Logger::ERROR,
+          "Fetching datafile from #{@datafile_url} failed. Error: #{e}"
+        )
+        return
+      end
+
+      handle_response response
+    end
+
+    def running
+      # Check if polling thread is alive or not.
+      @polling_thread.alive?
+    end
+
+    def run
+      # Triggered as part of the thread which fetches the datafile and sleeps until next update interval.
+      while running
+        fetch_datafile
+        sleep @update_interval
```

Comment:
Please add a check if we don't get response before our next schduled call, then no need to fetch again and as soon as get the response, then fetch datafile again immdiately.